Sources/ComposableArchitecture/Store.swift

```diff
@@ -348,21 +348,25 @@ public final class Store<State, Action> {
         guard !isSending else { return }
 
         let callbackInfo = Instrumentation.CallbackInfo<Self.Type, Any>(storeKind: Self.self, action: nil).eraseToAny()
-        instrumentation.callback?(callbackInfo, .pre, .storeToLocal)
+        instrumentation.callback?(callbackInfo, .pre, .scopedStoreToLocal)
         let newLocalState = toLocalState(newValue)
-        instrumentation.callback?(callbackInfo, .post, .storeToLocal)
+        instrumentation.callback?(callbackInfo, .post, .scopedStoreToLocal)
 
         guard let previousState = localStore?.state.value, let isDuplicate = isDuplicate else {
+            instrumentation.callback?(callbackInfo, .pre, .scopedStoreChangeState)
             localStore?.state.value = newLocalState
+            instrumentation.callback?(callbackInfo, .post, .scopedStoreChangeState)
```

Comment:
Maybe something like below would work?

```swift
{
  let inst = instrumentation.callback?(callbackInfo, .scopedStoreChangeState)
  localStore?.state.value = newLocalState
}()
```

This would require rejiggering the implementation a bit to wrap the consumer provided `callback` closure to return this framework object type. But that doesn't seem _impossible_ or even unreasonable. Not sure if others feel worth it?

I'm not entirely clear how well that works with the `defer` uses. Those depend on the specific ordering being what it is and I'm not sure deallocation would do that for us without further introducing these kind of wrapping closure scopes (or `do` blocks).

Reply:
I'm a little lost as to why would you want to do that tbh, I did use lifetime based tracking few times in my projects but when I was feeding that into normal codebase, if this is so isolated adding additional types and relying on cleanup logic adds complexity and extra memory frames ?

Reply:
Why not just add helper like this, I thought it's what you suggested in your previous comment:

```
class ScopedSignpost {
  private let instrumentation: Instrumentation
  private let callbackInfo: CallbackInfo
  private let signpost: CallbackKind

  init(instrumentation: Instrumentation, callbackInfo: CallbackInfo, signpost: CallbackKind) {
    this.instrumentation = instrumentation
    this.callbackInfo = callbackInfo
    this.signpost = signpost
    instrumentation.callback?(callbackInfo, .pre, signpost)
  }

  deinit {
     instrumentation.callback?(callbackInfo, .post, signpost)
  }
}
```

With that helper available a caller can decide to do one of three things:
 1. Use it directly as is because what they want to signpost actually is linine with existing scope.
 2. Add `do` block and use it if they think this improves structure and readability of the code.
 3. Not use helper at all, just use the standard API if they feel like the complication of adding extra scope is not worth it.

Reply:
Oh, I just noticed you pushed PR with prototype of that too and you actually did use something like that as underlaying mechanism. The only thing I kinda struggle with is why introduce the `callAsFunction` helper.
Between:
1. `callAsFunction` helper
```
        let newLocalState: LocalState = {
          let inst = instrumentation(callbackInfo, .scopedStoreToLocal)
          return toLocalState(newValue)
        } ()
```
2. Using the `ScopedSignpost` helper class directly
```
       let newLocalState: LocalState = {
         let signpost = ScopedSignpost(instrumentation, callbackInfo, .scopedStoreToLocal)
         retrun toLocalState(newValue)
      } ()
```
3. Using an a standard method as a builder of `ScopedSignpost`
```
       let newLocalState: LocalState = {
         let signpost = instrumentation.scopedSignpost(callbackInfo, .scopedStoreToLocal)
         retrun toLocalState(newValue)
      } ()
```

I would argue that (1) is the least readable as the semantic of calling the instrumentation as a function is rather unobvious. OTOH I think both (2) and (3) are expressing well that we are dealing with RAII.

Reply:
Well as it turns out Swift does not support this scoped RAII kind of usage. I knew it was a maybe due to reference counting, but figured worth a shot. @fischman-bcny provided the references proving this isn't a real path forward with Swift. 

Also, I kind like the call as function feature available in Swift. It isn't necessary but can be nice.